src/app/clusters/ecosystem-information-server/ecosystem-information-server.cpp

```diff
@@ -0,0 +1,368 @@
+{
+    Structs::EcosystemLocationStruct::Type locationStruct;
+    VerifyOrDie(!aUniqueLocationId.empty());
+    locationStruct.uniqueLocationID     = CharSpan(aUniqueLocationId.c_str(), aUniqueLocationId.size());
+    locationStruct.homeLocation         = GetEncodableHomeLocationStruct(mHomeLocation);
+    locationStruct.homeLocationLastEdit = mHomeLocationLastEditEpochUs;
+
+    // TODO(#33223) this is a hack, use mFabricIndex when it exists. Additionally checking fabric
+    // index matches should happen at the top of this method to prevent building the encodable
+    // device struct when we know it is not intended for the accessing fabric.
+    locationStruct.SetFabricIndex(aFabricIndex);
+    return aEncoder.Encode(locationStruct);
+}
+
+EcosystemInformationServer EcosystemInformationServer::mInstance;
+
+EcosystemInformationServer & EcosystemInformationServer::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR EcosystemInformationServer::AddDeviceInfo(EndpointId aEndpoint, std::unique_ptr<EcosystemDeviceStruct> aDevice)
+{
+    VerifyOrReturnError(aDevice, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError((aEndpoint != kRootEndpointId && aEndpoint != kInvalidEndpointId), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto & deviceInfo = mDevicesMap[aEndpoint];
+    VerifyOrReturnError((deviceInfo.mDeviceDirectory.size() >= kDeviceDirectoryMaxSize), CHIP_ERROR_NO_MEMORY);
+    deviceInfo.mDeviceDirectory.push_back(std::move(aDevice));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EcosystemInformationServer::AddLocationInfo(EndpointId aEndpoint, const std::string & aLocationId,
+                                                       std::unique_ptr<EcosystemLocationStruct> aLocation)
+{
+    VerifyOrReturnError(aLocation, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError((aEndpoint != kRootEndpointId && aEndpoint != kInvalidEndpointId), CHIP_ERROR_INVALID_ARGUMENT);
+
+    auto & deviceInfo = mDevicesMap[aEndpoint];
+    VerifyOrReturnError((deviceInfo.mLocationDirectory.find(aLocationId) == deviceInfo.mLocationDirectory.end()),
+                        CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError((deviceInfo.mLocationDirectory.size() >= kLocationDirectoryMaxSize), CHIP_ERROR_NO_MEMORY);
+    deviceInfo.mLocationDirectory[aLocationId] = std::move(aLocation);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EcosystemInformationServer::RemoveDevice(EndpointId aEndpoint, uint64_t aEpochUs)
+{
+    VerifyOrReturnError((mDevicesMap.find(aEndpoint) != mDevicesMap.end()), CHIP_ERROR_INVALID_ARGUMENT);
+    auto & deviceInfo = mDevicesMap[aEndpoint];
+    deviceInfo.mRemovedOn.SetValue(aEpochUs);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EcosystemInformationServer::EncodeRemovedOnAttribute(EndpointId aEndpoint, AttributeValueEncoder & aEncoder)
+{
+    if (mDevicesMap.find(aEndpoint) == mDevicesMap.end())
+    {
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedCluster);
```

Comment:
UnsupportedEndpoint?

or is the unsupported cluster correct? This may warrant a comment there since I read this as "search by endpoint and if not found return ..."

Reply:
I guess if mDevicesMap does not include a endpoint this means the cluster implementation does not exist on that endpoint so we always assume "endpoint is valid, but cluster may not be active there"?

Reply:
That is exactly the reason.